SOSS/dms/engine_utils.py

```diff
@@ -0,0 +1,1920 @@
+        return oversample_grid(out, n_os=n_os)
+
+
+def get_soss_grid(wave_maps, apertures, wave_min=0.55, wave_max=3.0, n_os=None):
+    # TODO replace wave_min, wave_max with wave_range?
+    """Create a wavelength grid specific to NIRISS SOSS mode observations.
+    Assumes 2 orders are given, use grid_from_map if only one order is needed.
+
+    Parameters
+    ----------
+    :param wave_maps: Array containing the pixel wavelengths for order 1 and 2.
+    :param apertures: Array containing the spatial profiles for order 1 and 2.
+    :param wave_min: Minimum wavelength the output grid should cover.
+    :param wave_max: Maximum wavelength the output grid should cover.
+    :param n_os: Oversampling of the grid compare to the pixel sampling. Can be
+        specified for each order if a list is given. If a single value is given
+        it will be used for all orders.
+
+    :type wave_maps: array[float]
+    :type apertures: array[float]
+    :type wave_min: float
+    :type wave_max: float
+    :type n_os: int or List[int]
+
+    :returns: wave_grid_soss - A wavelength grid optimized for extracting SOSS
+        spectra across order 1 and order 2.
+    :rtype: array[float]
+    """
+
+    # Check n_os input, default value is 2 for all orders.
+    if n_os is None:  # TODO check for integer type?
+        n_os = [2, 2]
+    elif np.ndim(n_os) == 0:
+        n_os = [n_os, n_os]
+    elif len(n_os) != 2:
+        msg = ("n_os must be an integer or a 2 element list or array of "
+               "integers, got {} instead")
+        raise ValueError(msg.format(n_os))
+
+    # Generate a wavelength range for each order.
+    # Order 1 covers the reddest part of the spectrum,
+    # so apply wave_max on order 1 and vice versa for order 2.
+
+    # Take the most restrictive wave_min for order 1
+    wave_min_o1 = np.maximum(wave_maps[0].min(), wave_min)
+
+    # Take the most restrictive wave_max for order 2.
+    wave_max_o2 = np.minimum(wave_maps[1].max(), wave_max)
+
+    # Now generate range for each orders
+    range_list = [[wave_min_o1, wave_max],
+                  [wave_min, wave_max_o2]]
+
+    # Use grid_from_map to construct separate oversampled grids for both orders.
+    wave_grid_o1 = grid_from_map(wave_maps[0], apertures[0],
+                                 wave_range=range_list[0], n_os=n_os[0])
+    wave_grid_o2 = grid_from_map(wave_maps[1], apertures[1],
+                                 wave_range=range_list[1], n_os=n_os[1])
+
+    # Keep only wavelengths in order 1 that aren't covered by order 2.
+    mask = wave_grid_o1 > wave_grid_o2.max()
+    wave_grid_o1 = wave_grid_o1[mask]
+
+    # Combine the order 1 and order 2 grids.
+    wave_grid_soss = np.concatenate([wave_grid_o1, wave_grid_o2])
+
+    # Sort values (and keep only unique).
+    wave_grid_soss = np.unique(wave_grid_soss)
+
+    return wave_grid_soss
+
+
+def _romberg_diff(b, c, k):
+    """Compute the differences for the Romberg quadrature corrections.
```

Comment:
I still don't like single letter variables

Reply:
This can be resolved at a later time if necessary.